### PR TITLE
LinearRegionSelector

### PR DESCRIPTION
**Pull request by kushalkolar**
starts #142

- [x] entire selection responds to mouse events
- [x] edges allow resizing with mouse events, edges change color and thickness with interaction
- [x] `LinearSelector.get_selected_data()` returns a view of the data, or a list of views if from a collection such as LineStack
- [x] `LinearBoundsFeature` manages events.
- [x] x-axis
- [x] y-axis
- [x] `LineGraphic.add_linear_region_selector()`, allows adding multiple selectors onto the same graphic
- [x] `LineCollection.add_linear_region_selector()`

As we make more selectors we can think about making a more centralized implementation for `Graphic.add_<selector>`. This may or may not be possible due to the diveristy of selectors that apply to a given graphic. For example, you wouldn't use a `LinearSelector` on a `ScatterGraphic`, and you wouldn't use a `RectangleSelector` on a `LineGraphic`.

```python
import fastplotlib as fpl
import numpy as np


gp = fpl.GridPlot((2, 2))

# preallocated size for zoomed data
zoomed_prealloc = 1_000

# data to plot
xs = np.linspace(0, 100, 1_000)
sine = np.sin(xs) * 20

# make sine along x axis
sine_graphic_x = gp[0, 0].add_line(sine)

# just something that looks different for line along y-axis
sine_y = sine
sine_y[sine_y > 0] = 0

# sine along y axis
sine_graphic_y = gp[0, 1].add_line(np.column_stack([sine_y, xs]))

# offset the position of the graphic to demonstrate `get_selected_data()` later
sine_graphic_y.position.set_x(50)
sine_graphic_y.position.set_y(50)

# add linear selectors
ls_x = sine_graphic_x.add_linear_region_selector()  # default axis is "x"
ls_y = sine_graphic_y.add_linear_region_selector(axis="y")

# preallocate array for storing zoomed in data
zoomed_init = np.column_stack([np.arange(zoomed_prealloc), np.random.rand(zoomed_prealloc)])

# make line graphics for displaying zoomed data
zoomed_x = gp[1, 0].add_line(zoomed_init)
zoomed_y = gp[1, 1].add_line(zoomed_init)


def interpolate(subdata: np.ndarray, axis: int):
    """1D interpolation to display within the preallocated data array"""
    x = np.arange(0, zoomed_prealloc)
    xp = np.linspace(0, zoomed_prealloc, subdata.shape[0])
    
    # interpolate to preallocated size
    return np.interp(x, xp, fp=subdata[:, axis])  # use the y-values


def set_zoom_x(ev):
    """sets zoomed x selector data"""
    selected_data = ev.pick_info["selected_data"]
    zoomed_x.data = interpolate(selected_data, axis=1)  # use the y-values
    gp[1, 0].auto_scale()


def set_zoom_y(ev):
    """sets zoomed y selector data"""
    selected_data = ev.pick_info["selected_data"]
    zoomed_y.data = -interpolate(selected_data, axis=0)  # use the x-values
    gp[1, 1].auto_scale()


# update zoomed plots when bounds change
ls_x.bounds.add_event_handler(set_zoom_x)
ls_y.bounds.add_event_handler(set_zoom_y)

gp.show()
```

https://user-images.githubusercontent.com/9403332/232291598-436c5af5-49c5-4a15-93aa-1c98278f133c.mp4

### CODE REVIEW

**Comment by kushalkolar** [comment]
@clewis7
@EricThomson

Some ideas on how to make these easy to use:

Just add a method to all graphics where it makes sense, like `LineGraphic` and `HeatmapGraphic`. It automatically sets the limits, height, etc. based on the the data in that graphic and adds it to the plot area!

```python
line = plot.add_line(...)

selector = line.add_linear_selector(resizable=True)
# similar syntax for heatmap
```

If they want full manual control for something more complex:

```python
from fastplotlib.graphics.selectors import LinearSelector
# do complex stuff
ls = LinearSelector(
    bounds, 
    limits, 
    height, 
    position,
    resizable=True
)
```

**Comment by EricThomson** [comment]
This looks great. I am not sure from reading just the parameter names what the difference between `bounds` and `limits` is, but it is clear from context (it is hard to name things like that I know from experience with windowing functions I've tried things like "view_range" and "selector_limits" to highlight that one is showing the range of the current viewpoint, the other is giving absolute limits of the selector, but honestly nothing will be perfect. That's what docs are for I guess. ¯\_(ツ)_/¯

I agree that creating abstractions so people can just attach these to data types so it works out of the box would be pretty amazing!

**Comment by kushalkolar** [comment]
More things to add to this and future selectors:

- Put a `parent_graphic` attribute on selector, can be `None` or can be a single graphic. In the future maybe we can do multiple graphics
  - have a `Selector.get_selected_<feature>()` where feature can be a `GraphicFeature` such as `data`, `colors`, `present` etc. and this returns the underlying feature data for that feature for the `parent_graphic`.

**Comment by kushalkolar** [comment]
Should allow this to be vertical as well so that it can be used with histograms for pyqtgraph-like vmin vmax setters

**Comment by kushalkolar** [comment]
I just realized it's useful to have a "selectors" property on LineGraphic (and the collections) which returns selectors attached to them. Just iterate through plot_area.graphics and return those which have "g.parent is self"

File: fastplotlib/plot.py [hunk 1]
```diff
@@ -100,6 +100,7 @@ def show(self):
 
         """
         self.canvas.request_draw(self.render)
-        self.auto_scale(maintain_aspect=True, zoom=0.95)
+        if autoscale:
+            self.auto_scale(maintain_aspect=True, zoom=0.95)
```
[the remark this answers is not in the document]
Comment (kushalkolar) [reply fastplotlib/plot.py:104]: This was here for a long time, it was just made conditional here

File: fastplotlib/graphics/selectors/linear.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        pick_info = {
+            "index": None,
+            "collection-index": self._collection_index,
+            "world_object": self._parent.world_object,
+            "new_data": new_data
+        }
+
+        event_data = FeatureEvent(type="bounds", pick_info=pick_info)
+
+        self._call_event_handlers(event_data)
```
[the remark this answers is not in the document]
Comment (kushalkolar) [reply fastplotlib/graphics/selectors/linear.py]: No but I will add it

File: fastplotlib/graphics/selectors/linear.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        Create a LinearSelector graphic which can be moved only along the x-axis. Useful for sub-selecting
+        data Line graphics or Heatmap graphics.
+
+        bounds[0], limits[0], and position[0] must be identical
+
+        Parameters
+        ----------
+        bounds: (int, int)
+            the initial bounds of the linear selector
+
```
[the remark this answers is not in the document]
Comment (kushalkolar) [reply fastplotlib/graphics/selectors/linear.py]: Yea

File: fastplotlib/graphics/selectors/_linear.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+            resizable: bool = False,
+            fill_color=(0, 0, 0.35),
+            edge_color=(0.8, 0.8, 0),
+            name: str = None
+    ):
+        """
+        Create a LinearSelector graphic which can be moved only along the x-axis. Useful for sub-selecting
+        data on Line graphics.
+
+        bounds[0], limits[0], and position[0] must be identical
```
[the remark this answers is not in the document]
Comment (kushalkolar) [reply fastplotlib/graphics/selectors/_linear.py:131]: This isn't the case anymore and I've remove this sanity check

File: fastplotlib/graphics/selectors/_linear.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+        Parameters
+        ----------
+        bounds: (int, int)
+            the initial bounds of the linear selector
+
+        limits: (int, int)
```
[the remark this answers is not in the document]
Comment (kushalkolar) [reply fastplotlib/graphics/selectors/_linear.py:138]: position is the initial xy position of the rectangular which is different from the bounds

File: fastplotlib/graphics/selectors/_linear.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+        if isinstance(source, GraphicCollection):
+            # this will return a list of views of the arrays, therefore no copy operations occur
+            # it's fine and fast even as a list of views because there is no re-allocating of memory
+            # this is fast even for slicing a 10,000 x 5,000 LineStack
+            return source[:].data[x_slice]
+
+        return source.data.buffer.data[x_slice]
+
+    def _add_plot_area_hook(self, plot_area):
```
[the remark this answers is not in the document]
Comment (kushalkolar) [reply fastplotlib/graphics/selectors/_linear.py:419]: Called when this gets added to a plot area

File: fastplotlib/graphics/selectors/_linear.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+        self._plot_area.controller.enabled = True
+
+    def _move_end(self, ev):
+        self._move_info = None
+        # sometimes weird stuff happens so we want to make sure the controller is reset
+        self._plot_area.controller.enabled = True
+
+        self._reset_edge_color()
+
+    def _pointer_enter_edge(self, edge: pygfx.Line, ev):
```
[the remark this answers is not in the document]
Comment (kushalkolar) [reply fastplotlib/graphics/selectors/_linear.py:536]: Highlights, changes color of the edge

File: fastplotlib/graphics/selectors/_linear.py [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+])
+
+y_bottom = np.array([
+    True, False,  True, False,  True, False,  True, False, False,
+    False, False, False,  True,  True,  True,  True,  True,  True,
+    False, False,  True,  True, False, False
+])
+
+
+class _LinearBoundsFeature(GraphicFeature):
```
[the remark this answers is not in the document]
Comment (kushalkolar) [reply fastplotlib/graphics/selectors/_linear.py]: This manages the bounds feature, changes the rectangle Geometry and edge lines, and handles events.

**Pull request merged**